### DISCUSSION TITLE
Trouble annotating a sometimes-nested structure

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/pyright#5214] [Q&A] [answered]
What's the proper way to annotate this (admittedly strange) function? I've tried a huge number of variations and nothing's worked for me. Below is the simplest thing I've tried.

```python
from typing import Dict, List, Union

NestedStrList = Union[str, List["NestedStrList"]]

def complex_interpreter(src: str) -> NestedStrList:
    nesting = src.count("[")
    result = src.replace("[", "")
    for _ in range(nesting):
        result = [result]
    return result

def main():
    import sys
    for arg in sys.argv[1:]:
        result = complex_interpreter(arg)

        import json
        print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()
```

```sh
$ python demo5.py 'x' '[y' '[[[z'
"x"
[
  "y"
]
[
  [
    [
      "z"
    ]
  ]
]

$ pyright demo5.py
./demo5.py
  ./demo5.py:11:12 - error: Expression of type "str | list[str | list[str | list[str]]]" cannot be assigned to return type "NestedStrList"
    Type "str | list[str | list[str | list[str]]]" cannot be assigned to type "NestedStrList"
      "list[str | list[str | list[str]]]" is incompatible with "List[NestedStrList]"
        TypeVar "_T@list" is invariant
          Type "str | list[str | list[str]]" cannot be assigned to type "NestedStrList"
            "list[str | list[str]]" is incompatible with "List[NestedStrList]"
            Type "list[str | list[str]]" cannot be assigned to type "NestedStrList"
      Type "list[str | list[str | list[str]]]" cannot be assigned to type "NestedStrList"
        "list[str | list[str | list[str]]]" is incompatible with "str"
    ... (reportGeneralTypeIssues)
  ./demo5.py:1:20 - error: Import "Dict" is not accessed (reportUnusedImport)
2 errors, 0 warnings, 0 informations
```

**Comment by erictraut** [comment, marked as answer]
You've annotated the function signature correctly. The problem is that you haven't annotated the local variable `result`, so standard type inference is being used for the expression `[result]`. If you declare the type of `result` as `NestedStrList`, then bidirectional type inference will be used when evaluating the expression `[result]`.

**Reply by ghost** [reply]
I swear I tried that ... but it works perfectly. Thanks again!

**Reply by ghost** [reply]
Actually, I did try that. This is a bit closer to real life, and that extra annotation doesn't do the trick:

```python
from typing import List, Union

NestedStrList = Union[str, List["NestedStrList"]]


def complex_interpreter(src: str) -> NestedStrList:
    nesting = src.count("[")
    result: NestedStrList = src.split("[")
    for _ in range(nesting):
        result = [result]
    return result
```

```sh
$ python  demo6.py '[[[3' '3'
[[[["", "", "", "3"]]]]
["3"]

$ pyright demo6.py
./demo6.py
  ./demo6.py:8:29 - error: Expression of type "list[str]" cannot be assigned to declared type "NestedStrList"
    Type "list[str]" cannot be assigned to type "NestedStrList"
      "list[str]" is incompatible with "str"
      "list[str]" is incompatible with "List[NestedStrList]"
        TypeVar "_T@list" is invariant
          Type "str" cannot be assigned to type "NestedStrList"
            "str" is incompatible with "List[NestedStrList]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
exit code: 1
```

**Reply by ghost** [reply]
This "works"

```python
NestedStrList = Union[str, Sequence["NestedStrList"]]
```

except it also somehow silently disable typechecking: (note the `int`)

```python
NestedIntList = Union[int, Sequence["NestedIntList"]]

def complex_interpreter(src: str) -> NestedIntList:
    nesting = src.count("[")
    result: NestedIntList = src.split("[")
    for _ in range(nesting):
        result = [result]
    return result
```


```sh
$ pyright demo6.py
0 errors, 0 warnings, 0 informations

$ mypy demo6.py
Success: no issues found in 1 source file

$ pytype demo6.py
Success: no errors found
```

**Reply by erictraut** [reply]
The extra annotation doesn't work with a call to `split` because this method returns a `list[str]`. That's not type compatible with `list[NestedStrList]` because the type parameter for `list` is invariant. You would need to construct a new list from the results returned by `split` if you want it to be compatible with the type `NestedStrList`.

```python
    result: NestedStrList = list(src.split("["))
```
or
```python
    result: NestedStrList = [x for x in src.split("[")]
```

The "silently disable typechecking" looks like a bug. If you want to open a bug report, I can take a look at it.

**Reply by ghost** [reply]
Here's the bug report: https://github.com/microsoft/pyright/issues/5215

I'm hearing that this is a feature and not a bug, but how do I rationalize this?
I understand that list[T] is invariant, and I think I understand why ( if it were covariant then a mutating function of `list[Iterable]` might leave you with a list of tuple where you expected a list of list).

What I can't wrap my head around though, is why is it not okay to return an invariant generic that is a subtype of a function's return value? (Is it bad terminology to say that a return value is contravariant with respect to the function's return type?) A return value is guaranteed to be a subclass of a return value, and is not (invariant) `list[tuple]` a subtype of `Sequence[Sequence]` ?

Oh! I (maybe) see the issue; the `TypeVar "_T@list"` it's referring to here is not the one in the  return *value* but in the *function's return type*,  and that's why returning a (notional) subclass isn't allowed ?

**Reply by erictraut** [reply]
> how do I rationalize this?

This behavior is required by the Python type system. The type `list[str]` is not a subtype of `NestedStrList`, so you cannot assign the former to the latter.

> I understand that list[T] is invariant

That statement is not quite correct. It would be correct to say "the type variable `T@list` is invariant". It doesn't make sense to say that `list[T]` is invariant. Variance is a property of a type parameter, not a class. There are classes (e.g. `Mapping`) that have multiple type parameters whose variance differs from each other. I realize that I'm being a bit pedantic here, but I'm hoping that helps clear up some confusion for you.

> why is it not okay to return an invariant generic that is a subtype of a function's return value?

It is OK to return a value whose type is a subtype of the declared return type of the function. But, as I noted above, `list[str]` is not a subtype of `NestedStrList`.

**Reply by ghost** [reply]
> I realize that I'm being a bit pedantic here ...

No I'm happy to learn the correct terminology :)

> It would be correct to say "the type variable T@list is invariant". It doesn't make sense to say that list[T] is invariant.

In all three articles below, the terminology used is "$constructor is $variant in $typevar`, as in "`list[T]` is invariant in `T`", which to my eye is equivalent but more familiar than `T@list is invariant` (since there's similar use of `@` in python).  I improperly elided the "in `T`" bit because it seemed unambiguous to me.

1. https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
2. https://mypy.readthedocs.io/en/stable/generics.html
3. https://docs.scala-lang.org/tour/variances.html#covariance -- actually seems to mix the two styles

But I do see that C#, and microsoft research consistently use your terminology, specifying the variable and not the constructor as variant.
